sklearn/impute/_base.py

```diff
@@ -38,6 +40,20 @@ def _check_inputs_dtype(X, missing_values):
         )
 
 
+def _safe_min(items):
+    """Compute the minimum of a list of potentially non-comparable values.
+    If values cannot be directly compared due to type incompatibility,
+    the object with the lowest hash of its pickled representation is returned
+    as a fallback."""
+    items = list(items)  # necessary if items is a generator
+    try:
+        return min(items)
+    except TypeError as e:
+        if "'<' not supported between" in str(e):
+            return min(items, key=lambda x: murmurhash3_32(pickle.dumps(x)))
```

Comment:
I wonder if it's not a bit overkill. For the use case of SimpleImputer, I believe that using the string representation of objects would be fine. It's deterministic across versions, os, envs... for usual types (int, float, str, None). It's not on more complex objects but I don't think that we expect complex objects to be passed to SimpleImputer.
```suggestion
            return min(items, key=str)
```
What do you think ?

Reply:
I thought of that too but I am bugged by the 1 out of a million case where there is a tie between `None` and `'None'` that is shuffle-sensitive:

```python
In [1]: min([None, 'None'], key=str)

In [2]: min(['None', None], key=str)
Out[2]: 'None'
```

`repr` does not fail in that case but it may in others.

Actually, we do not really need the hash, the pickle dumped string should be unique.
```python
            return min(items, key=lambda x: pickle.dumps(x))
```
Is that better? Still needs pickle though... 
Actually, that's where the other PR comes in handy, using the type breaks the tie.

Reply:
It's actually the same for `1` and `"1"` and so on which is more likely to happen.

I was undecided between `pickle.dumps(x)` and `(str(type(x), str(x))`, but a quick benchmark showed that the latter is much faster, so I'm in favor of choosing this one.